src/epgsql_sock.erl

```diff
@@ -89,13 +91,20 @@
                 repl_feedback_required,
                 repl_cbmodule,
                 repl_cbstate,
-                repl_receiver}).
+                repl_receiver,
+                timer}).
```

Comment:
UPD: I made some investigation and found following things:

* `epgsql:cancel/1` cancels only ONE request from pipeline (which is currently running one)
* connection is still usable after cancel

```erlang
2> {ok, C} = epgsql:connect(#{host => "localhost", port => 5433, username => "wgm", password=>"wgm", database=>"wgm1"}).
{ok,<0.142.0>}
% 3 commands are put in pipeline
19> epgsqla:squery(C, <<"SELECT pg_sleep(100)">>).
#Ref<0.0.1.1136>
20> epgsqla:squery(C, <<"SELECT pg_sleep(100)">>).
#Ref<0.0.1.1141>
21> epgsqla:squery(C, <<"SELECT pg_sleep(100)">>).
#Ref<0.0.1.1146>
% cancel 1st 2 of them
22> epgsqla:cancel(C).                            
ok
23> epgsqla:cancel(C).
ok
% got 2 'query_cancelled' responses (please note references order: #Ref<0.0.1.1136> then #Ref<0.0.1.1141> then #Ref<0.0.1.1146>)
24> flush().                                      
Shell got {<0.142.0>,#Ref<0.0.1.1136>,
           {error,{error,error,<<"57014">>,query_canceled,
                         <<"canceling statement due to user request">>,
                         [{file,<<"postgres.c">>},
                          {line,<<"3003">>},
                          {routine,<<"ProcessInterrupts">>}]}}}
Shell got {<0.142.0>,#Ref<0.0.1.1141>,
           {error,{error,error,<<"57014">>,query_canceled,
                         <<"canceling statement due to user request">>,
                         [{file,<<"postgres.c">>},
                          {line,<<"3003">>},
                          {routine,<<"ProcessInterrupts">>}]}}}
ok
% cancel 3rd request
25> epgsqla:cancel(C).
ok
% got 3rd 'query cancelled'
26> flush().          
Shell got {<0.142.0>,#Ref<0.0.1.1146>,
           {error,{error,error,<<"57014">>,query_canceled,
                         <<"canceling statement due to user request">>,
                         [{file,<<"postgres.c">>},
                          {line,<<"3003">>},
                          {routine,<<"ProcessInterrupts">>}]}}}
ok
% following cancels are noops
27> epgsqla:cancel(C).
ok
28> flush().          
ok
% connection is still usable
29> epgsql:squery(C, <<"SELECT pg_sleep(1)">>).   
{ok,[{column,<<"pg_sleep">>,void,4,-1,0}],[{<<>>}]}
```

It means that if we have 2 requests in a pipeline with different timeouts:

```
epgsqla:timed_squery(C, "SELECT pg_sleep(10)", 15000)
epgsqla:timed_squery(C, "SELECT pg_sleep(10)", 1000)
```
We expect 1st one to terminate successfully but the 2nd one with error. But we'll get absolutely different result: 1st query will be cancelled by 2nd timer.

I don't see any easy solution for this, so, we, probably, may just leave your curent implementation for now, but don't forget to document this behaviour of async interface. Pipelining is not so widely used feature, but it adds a bit of headache...